src/goto-programs/goto_clean_expr.cpp

```diff
@@ -429,6 +438,22 @@ void goto_convertt::clean_expr(
       expr.operands().size() == 1, "ID_compound_literal has a single operand");
     expr = to_unary_expr(expr).op();
   }
+  else if(expr.id() == ID_count_leading_zeros)
+  {
+    count_leading_zeros_exprt &clz = to_count_leading_zeros_expr(expr);
+    if(!clz.zero_permitted())
+    {
+      clz.zero_permitted(true);
+      const source_locationt source_location = clz.source_location();
+
```

Comment:
I would go stronger than the nondet: The wording in the gcc docs is that __builtin_clz is undefined when the argument is zero. Assuming they use the meaning of "undefined" as in the ISO C standard, then this is worth an assertion, to be generated in goto_check.

Reply:
Hmm, true, then I could perhaps demote the `ID_zero_permitted` to `ID_C_zero_permitted`?

Reply:
I don't think you need it at all. We don't have anything equivalent for any of the other expressions that are only partially defined.

Reply:
But we need to distinguish whether we're handling MSVC's `__lzcnt` or `__builtin_clz`. The former is defined for all values, the latter isn't defined for `0`.

Reply:
Ah, right. Then I'd make it `ID_zero_permitted`; a comment would be too weak, given that the semantics do differ.

Reply:
Except they only sort-of do: it seems that in practice it's almost always ok, it's just that GCC doesn't provide a guarantee: https://stackoverflow.com/questions/19527897/how-undefined-are-builtin-ctz0-or-builtin-clz0. How about the solution that I just pushed?